File: frontend/src/lib/currency.ts
/**
 * Currency utilities and constants
 */

export interface CurrencyInfo {
  code: string;
  symbol: string;
  name: string;
  locale: string;
  decimalPlaces: number;
  thousandsSeparator: string;
  decimalSeparator: string;
  grouping: number;
}

export const CURRENCIES: CurrencyInfo[] = [
  {
    code: "USD",
    symbol: "$",
    name: "US Dollar",
    locale: "en-US",
    decimalPlaces: 2,
    thousandsSeparator: ",",
    decimalSeparator: ".",
    grouping: 3,
  },
  {
    code: "COP",
    symbol: "$",
    name: "Colombian Peso",
    locale: "es-CO",
    decimalPlaces: 0, // COP typically doesn't use decimals
    thousandsSeparator: ".",
    decimalSeparator: ",",
    grouping: 3,
  },
  {
    code: "ARS",
    symbol: "$",
    name: "Argentine Peso",
    locale: "es-AR",
    decimalPlaces: 2,
    thousandsSeparator: ".",
    decimalSeparator: ",",
    grouping: 3,
  },
  {
    code: "EUR",
    symbol: "€",
    name: "Euro",
    locale: "en-EU",
    decimalPlaces: 2,
    thousandsSeparator: ".",
    decimalSeparator: ",",
    grouping: 3,
  },
] as const;

export const SUPPORTED_CURRENCIES = CURRENCIES;

export type CurrencyCode = typeof CURRENCIES[number]["code"];

export function getCurrencySymbol(currency: string): string {
  return CURRENCIES.find((c) => c.code === currency)?.symbol || "$";
}

export function getCurrencyName(currency: string): string {
  return CURRENCIES.find((c) => c.code === currency)?.name || "US Dollar";
}

export function getCurrencyInfo(currency: string): CurrencyInfo {
  return CURRENCIES.find((c) => c.code === currency) || CURRENCIES[0];
}

/**
 * Format currency amount with proper thousands/millions grouping
 * Example: COP 1.000.000, USD 1,000,000.00
 */
export function formatCurrency(
  amount: number,
  currency: string = "USD",
  useGrouping: boolean = true
): string {
  // #region agent log
  fetch('http://127.0.0.1:7244/ingest/9259ea1e-d9d4-4580-890f-411d9fb62b18',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify({location:'lib/currency.ts:79',message:'formatCurrency called',data:{amount,amountType:typeof amount,isNaN:isNaN(amount),currency},timestamp:Date.now(),sessionId:'debug-session',runId:'run1',hypothesisId:'C'})}).catch(()=>{});
  // #endregion
  
  // Handle NaN and invalid numbers
  if (isNaN(amount) || !isFinite(amount)) {
    // #region agent log
    fetch('http://127.0.0.1:7244/ingest/9259ea1e-d9d4-4580-890f-411d9fb62b18',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify({location:'lib/currency.ts:85',message:'formatCurrency received NaN or invalid number',data:{amount,amountType:typeof amount},timestamp:Date.now(),sessionId:'debug-session',runId:'run1',hypothesisId:'C'})}).catch(()=>{});
    // #endregion
    return `${getCurrencyInfo(currency).symbol} 0`;
  }
  
  const currencyInfo = getCurrencyInfo(currency);
  const symbol = currencyInfo.symbol;
  const decimalPlaces = currencyInfo.decimalPlaces;
  const thousandsSep = currencyInfo.thousandsSeparator;
  const decimalSep = currencyInfo.decimalSeparator;
  const grouping = currencyInfo.grouping;

  // Round to appropriate decimal places
  const roundedAmount = Math.round(amount * Math.pow(10, decimalPlaces)) / Math.pow(10, decimalPlaces);

  // Split into integer and decimal parts
  const integerPart = Math.floor(Math.abs(roundedAmount));
  const decimalPart = Math.abs(roundedAmount) - integerPart;

  // Format integer part with grouping
  let integerStr: string;
  if (useGrouping && grouping > 0) {
    const reversedStr = integerPart.toString().split("").reverse().join("");
    const groupedParts: string[] = [];
    for (let i = 0; i < reversedStr.length; i += grouping) {
      groupedParts.push(reversedStr.slice(i, i + grouping));
    }
    integerStr = groupedParts.join(thousandsSep).split("").reverse().join("");
  } else {
    integerStr = integerPart.toString();
  }

  // Format decimal part
  let formattedAmount: string;
  if (decimalPlaces > 0 && decimalPart > 0) {
    const decimalStr = decimalPart.toFixed(decimalPlaces).slice(2); // Remove "0."
    formattedAmount = `${integerStr}${decimalSep}${decimalStr}`;
  } else {
    formattedAmount = integerStr;
  }

  // Add negative sign if needed
  if (roundedAmount < 0) {
    formattedAmount = `-${formattedAmount}`;
  }

  // Add currency symbol
  return `${symbol} ${formattedAmount}`;
}

File: backend/app/core/currency.py
"""
Currency utilities and constants
ESTÁNDAR NOUGRAM: Usa Money para precisión en formateo y conversiones
"""
import math
from decimal import Decimal
from enum import Enum
from typing import Any

from app.core.money import Money


class Currency(str, Enum):
    """Supported currencies"""
    USD = "USD"  # US Dollar
    COP = "COP"  # Colombian Peso
    ARS = "ARS"  # Argentine Peso
    EUR = "EUR"  # Euro


CURRENCY_INFO: dict[str, dict[str, Any]] = {
    "USD": {
        "symbol": "$",
        "name": "US Dollar",
        "locale": "en-US",
        "decimal_places": 2,
        "thousands_separator": ",",  # Thousands separator
        "decimal_separator": ".",  # Decimal separator
        "grouping": 3,  # Group digits by 3 (thousands, millions, etc.)
    },
    "COP": {
        "symbol": "$",
        "name": "Colombian Peso",
        "locale": "es-CO",
        "decimal_places": 0,  # COP typically doesn't use decimals
        "thousands_separator": ".",  # Thousands separator (period in Spanish)
        "decimal_separator": ",",  # Decimal separator (comma in Spanish)
        "grouping": 3,  # Group digits by 3 (thousands, millions, etc.)
    },
    "ARS": {
        "symbol": "$",
        "name": "Argentine Peso",
        "locale": "es-AR",
        "decimal_places": 2,
        "thousands_separator": ".",  # Thousands separator
        "decimal_separator": ",",  # Decimal separator
        "grouping": 3,  # Group digits by 3 (thousands, millions, etc.)
    },
    "EUR": {
        "symbol": "€",
        "name": "Euro",
        "locale": "en-EU",
        "decimal_places": 2,
        "thousands_separator": ".",  # Thousands separator
        "decimal_separator": ",",  # Decimal separator
        "grouping": 3,  # Group digits by 3 (thousands, millions, etc.)
    },
}

# Exchange rates to USD (base currency)
# NOTE: These are example rates. In production, you should fetch real-time rates
# from an API like exchangerate-api.com, fixer.io, or similar
EXCHANGE_RATES_TO_USD: dict[str, float] = {
    "USD": 1.0,      # Base currency
    "COP": 4000.0,   # Example: 1 USD = 4000 COP (update with real rates)
    "ARS": 850.0,    # Example: 1 USD = 850 ARS (update with real rates)
    "EUR": 0.92,     # Example: 1 USD = 0.92 EUR (update with real rates)
}


def format_currency(
    amount: float | Decimal | Money,
    currency: str = "USD",
    use_grouping: bool = True
) -> str:
    """
    Format amount as currency string with proper thousands/millions grouping
    ESTÁNDAR NOUGRAM: Acepta Money, Decimal o float para compatibilidad
    
    Args:
        amount: Amount to format (Money, Decimal, or float)
        currency: Currency code (USD, COP, ARS, EUR). Si amount es Money, se usa su currency
        use_grouping: Whether to use thousands/millions grouping (default: True)
        
    Returns:
        Formatted currency string (e.g., "$ 1.000.000" for COP, "$ 1,000,000.00" for USD)
    """
    # ESTÁNDAR NOUGRAM: Si es Money, usar su currency y quantizar
    if isinstance(amount, Money):
        currency = amount.currency
        # Quantizar según la moneda
        if currency == "COP":
            quantized = amount.quantize()
        else:
            quantized = amount.quantize()
        amount_decimal = quantized.amount
    elif isinstance(amount, Decimal):
        amount_decimal = amount
    else:
        # Compatibilidad hacia atrás: convertir float a Decimal
        amount_decimal = Decimal(str(amount))
    
    currency_info = CURRENCY_INFO.get(currency, CURRENCY_INFO["USD"])
    symbol = currency_info.get("symbol", "$")
    decimal_places = currency_info.get("decimal_places", 2)
    thousands_sep = currency_info.get("thousands_separator", ",")
    decimal_sep = currency_info.get("decimal_separator", ".")
    grouping = currency_info.get("grouping", 3)
    
    # ESTÁNDAR NOUGRAM: Redondear usando Decimal para precisión
    precision = Decimal('1') if decimal_places == 0 else Decimal('0.1') ** decimal_places
    rounded_decimal = amount_decimal.quantize(precision)
    
    # Convertir a float solo para formateo (ya está redondeado)
    rounded_amount = float(rounded_decimal)
    
    # Split into integer and decimal parts
    integer_part = int(abs(rounded_amount))
    decimal_part = abs(rounded_amount) - integer_part
    
    # Format integer part with grouping
    if use_grouping and grouping > 0:
        integer_str = str(integer_part)
        # Reverse string to group from right to left
        reversed_str = integer_str[::-1]
        grouped_parts = []
        for i in range(0, len(reversed_str), grouping):
            grouped_parts.append(reversed_str[i:i+grouping])
        grouped_str = thousands_sep.join(grouped_parts)
        integer_str = grouped_str[::-1]
    else:
        integer_str = str(integer_part)
    
    # Format decimal part
    if decimal_places > 0 and decimal_part > 0:
        decimal_str = f"{decimal_part:.{decimal_places}f}"[2:]  # Remove "0."
        formatted_amount = f"{integer_str}{decimal_sep}{decimal_str}"
    else:
        formatted_amount = integer_str
    
    # Add negative sign if needed
    if rounded_amount < 0:
        formatted_amount = f"-{formatted_amount}"
    
    # Add currency symbol
    return f"{symbol} {formatted_amount}"


def get_currency_symbol(currency: str = "USD") -> str:
    """
    Get currency symbol
    
    Args:
        currency: Currency code
        
    Returns:
        Currency symbol
    """
    return CURRENCY_INFO.get(currency, CURRENCY_INFO["USD"]).get("symbol", "$")


def get_currency_name(currency: str = "USD") -> str:
    """
    Get currency name
    
    Args:
        currency: Currency code
        
    Returns:
        Currency name
    """
    return CURRENCY_INFO.get(currency, CURRENCY_INFO["USD"]).get("name", "US Dollar")


def is_valid_currency(currency: str) -> bool:
    """
    Check if currency code is valid
    
    Args:
        currency: Currency code to validate
        
    Returns:
        True if valid, False otherwise
    """
    return currency in CURRENCY_INFO


def get_all_currencies() -> list[dict[str, str]]:
    """
    Get list of all supported currencies
    
    Returns:
        List of currency dictionaries with code, symbol, and name
    """
    return [
        {
            "code": code,
            "symbol": info["symbol"],
            "name": info["name"],
        }
        for code, info in CURRENCY_INFO.items()
    ]


def convert_currency(
    amount: float | Decimal | Money,
    from_currency: str,
    to_currency: str
) -> float | Money:
    """
    Convert amount from one currency to another
    ESTÁNDAR NOUGRAM: Usa Money para precisión en conversiones
    
    Args:
        amount: Amount to convert (Money, Decimal, or float)
        from_currency: Source currency code. Si amount es Money, se usa su currency
        to_currency: Target currency code
        
    Returns:
        Converted amount (Money si input es Money, float para compatibilidad)
    """
    # ESTÁNDAR NOUGRAM: Si es Money, usar su currency
    is_money = isinstance(amount, Money)
    if is_money:
        from_currency = amount.currency
        amount_decimal = amount.amount
    elif isinstance(amount, Decimal):
        amount_decimal = amount
    else:
        amount_decimal = Decimal(str(amount))
    
    if from_currency == to_currency:
        if is_money:
            return amount
        return float(amount_decimal)
    
    # ESTÁNDAR NOUGRAM: Convertir usando Decimal para precisión
    # Convert to USD first (base currency)
    from_rate = Decimal(str(EXCHANGE_RATES_TO_USD.get(from_currency, 1.0)))
    amount_in_usd = amount_decimal / from_rate
    
    # Convert from USD to target currency
    to_rate = Decimal(str(EXCHANGE_RATES_TO_USD.get(to_currency, 1.0)))
    converted_decimal = amount_in_usd * to_rate
    
    # Retornar Money si input era Money, float para compatibilidad
    if is_money:
        return Money(converted_decimal, to_currency)
    return float(converted_decimal)


def normalize_to_primary_currency(
    amount: float | Decimal | Money,
    from_currency: str | None,
    primary_currency: str = "USD"
) -> float | Money:
    """
    Normalize amount to primary currency
    ESTÁNDAR NOUGRAM: Usa Money para precisión en normalización
    
    This is used for calculations where all costs need to be in the same currency.
    
    Args:
        amount: Amount to normalize (Money, Decimal, or float)
        from_currency: Source currency code (can be None, defaults to USD). Si amount es Money, se ignora
        primary_currency: Target primary currency code (default: USD)
        
    Returns:
        Normalized amount in primary currency (Money si input es Money, float para compatibilidad)
    """
    # ESTÁNDAR NOUGRAM: Si es Money, usar su currency
    is_money = isinstance(amount, Money)
    if is_money:
        from_currency = amount.currency
    
    # Handle None or empty currency
    if not from_currency or from_currency not in CURRENCY_INFO:
        from_currency = "USD"
    
    if not primary_currency or primary_currency not in CURRENCY_INFO:
        primary_currency = "USD"
    
    return convert_currency(amount, from_currency, primary_currency)
